src/vs/base/browser/ui/iconLabel/iconLabelHover.ts

```diff
@@ -35,18 +35,94 @@ export interface ICustomHover extends IDisposable {
 	 * Allows to programmatically hide the hover.
 	 */
 	hide(): void;
+
+	/**
+	 * Updates the contents of the hover.
+	 */
+	update(tooltip: string | IIconLabelMarkdownString | HTMLElement): void;
 }
 
-export function setupCustomHover(hoverDelegate: IHoverDelegate, htmlElement: HTMLElement, markdownTooltip: string | IIconLabelMarkdownString | HTMLElement | undefined): ICustomHover | undefined {
-	if (!markdownTooltip) {
-		return undefined;
+type MarkdownTooltipContent = string | IIconLabelMarkdownString | HTMLElement | undefined;
+type ResolvedMarkdownTooltipContent = IMarkdownString | string | HTMLElement | undefined;
+class UpdatableHoverWidget implements IDisposable {
+
+	private _hoverWidget: IHoverWidget | undefined;
+	private _cancellationTokenSource: CancellationTokenSource | undefined;
+
+	constructor(private hoverDelegate: IHoverDelegate, private target: IHoverDelegateTarget, private fadeInAnimation: boolean) {
+	}
+
+	async update(markdownTooltip: MarkdownTooltipContent): Promise<void> {
+		if (this._cancellationTokenSource) {
+			// there's an computation ongoing, cancel it
+			this._cancellationTokenSource.dispose(true);
+			this._cancellationTokenSource = undefined;
+		}
+		if (this.isDisposed) {
+			return;
+		}
+
+		let resolvedContent;
+		if (markdownTooltip === undefined || isString(markdownTooltip) || markdownTooltip instanceof HTMLElement) {
+			resolvedContent = markdownTooltip;
+		} else if (!isFunction(markdownTooltip.markdown)) {
+			resolvedContent = markdownTooltip.markdown ?? markdownTooltip.markdownNotSupportedFallback;
+		} else {
+			// compute the content, potentially long-running
+
+			// show 'Loading' if no hover is up yet
+			if (!this._hoverWidget) {
+				this.show(localize('iconLabel.loading', "Loading..."));
+			}
+
+			// compute the content
+			const cancellationTokenSource = this._cancellationTokenSource = new CancellationTokenSource();
+			resolvedContent = await markdownTooltip.markdown(cancellationTokenSource.token);
+
+			if (this.isDisposed || cancellationTokenSource !== this._cancellationTokenSource) {
```

Comment:
I find the `cancellationTokenSource !== this._cancellationTokenSource` slightly weird, couldn't you just check if the token is canceled? Because the token source is only different when being disposed.

Reply:
Pushed via https://github.com/microsoft/vscode/commit/6583c2b0ef0ece70a1a2f2cd09681e3e29159c85